pkg/buildid/file.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package buildid
+
+import (
+	"bufio"
+	"bytes"
+	"debug/elf"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"strconv"
+)
+
+var elfPrefix = []byte("\x7fELF")
+
+var readSize = 32 * 1024 // changed for testing
+
+// FromFile reads the build ID from a binary by opening the file,
+// but only reading the first 32 kB of the file.
+// It also treats Go binaries specially, reading the build ID from
+// designated sections.
+// ELF binaries store the build ID in a proper PT_NOTE section.
+func FromFile(f *os.File) (id string, err error) { //nolint:nonamedreturns
+	// Read the first 32 kB of the binary file.
+	// That should be enough to find the build ID.
+	// In ELF files, the build ID is in the leading headers,
+	// which are typically less than 4 kB, not to mention 32 kB.
+	data := make([]byte, readSize)
+	_, err = io.ReadFull(bufio.NewReader(f), data)
+	if errors.Is(err, io.ErrUnexpectedEOF) {
+		err = nil
+	}
+	if err != nil {
+		return "", err
+	}
+
+	if bytes.HasPrefix(data, elfPrefix) {
+		defer func() {
```

Comment:
We’re only reading 32kb, why not just use readat? Any seeking around feels very error prone.

Reply:
Good point. Will do.